### ISSUE TITLE
[wave] llama4 operations

### ISSUE THREAD

**Issue by ftynse** [iree-org/iree-turbine#1042]
- Transformer (main model)
	- VisionEmbeddings (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/model.py#L367)
		- Reshape (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/embedding.py#L171-L172)
		- VisionEncoder
			- More reshapes
			- Conv1D (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/encoder.py#L389)
			- Class embedding (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/encoder.py#L394)
				- concat (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/encoder.py#L367-L376)
			- add (elementwise) for positional embedding
			- LayerNorm (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/encoder.py#L256-L257)
			- Transformer (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/encoder.py#L259-L267) – this is similar to the main transformer, **NOT** detailing too much
				- GeLU
			- trailing element removal (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/encoder.py#L419-L420)
			- reshape + concat (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/encoder.py#L426)
		- VisionAdapter
			- PixelShuffleMLP (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/embedding.py#L129-L132)
				- PixelShuffle
					- rehsapes (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/embedding.py#L30-L32)
					- in particular, pixel shuffle op (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/embedding.py#L36-L47)
				- MLP
					- ColumnParallelLinear (matmul transpose) (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/embedding.py#L61)
					- RowParallelLinear (matmul) (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/embedding.py#L67)
					- GELU
					- dropout (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/embedding.py#L79)
				- ColumnParallelLinear (matmul transpose)
	- projection: ColumnParallelLinear (matmul transpose) + mask (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/model.py#L401-L402)
	- scatter embeddings (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/embedding.py#L177)
		- masked scatter (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/vision/embedding.py#L205-L208)
	- layer: TransformerBlock
		- Attention
			- triangular masking, aka causal attention, (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/model.py#L408-L419)
			- chunked attention (variation of the above for iRoPe)
			- Rotary embedding, presumably _inactive_ in LLama 4 (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/model.py#L89-L99), there is another way to implement this via a 2x2 rotation matrix without needing support for complex values
			- RMSNorm, optional (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/model.py#L216-L218)
			- Temperature tuning (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/model.py#L223-L229)
				- arange/iota
				- log (elementwise)
				- floor (elementwise)
				- add (elementwise)
				- divide (elementwise)
				- mul (elementwise)
			- Matmul/matvec (prefill/decode)
			- Softmax
		- add (elementwise)
		- FeedForward (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/model.py#L280-L282)
		- MoE (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/model.py#L266-L271)
		- RMSNorm (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/model.py#L285-L286)
	- RMSNorm (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/model.py#L423)
	- ColumnParallelLinear (matmul transpose) (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/model.py#L424)
- Full token generation (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/generation.py#L207-L211):
	- softmax
	- argmax
	- sample_top_p (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/generation.py#L292)
		- sort
		- cumsum
		- sum (reduction)
		- div (elementwise)
		- multinomial (sampling from distribution, needs RNG)
		- gather
	- cross_entropy (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/generation.py#L220 see https://docs.pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss for definition)
		- exp (elementwise)
		- sum/mean (reduction)
		- div (elementwise)
		- log (elementwise)
		- Note: can be seen as generalization of softmax
- NOT PRESENT in the model, but may be present in variations`logits_processor` (https://github.com/meta-llama/llama-models/blob/01dc8ce46fecf06b639598f715efbb4ab981fb4c/models/llama4/generation.py#L205)
